### ISSUE TITLE
Attention softmax is applied to incorrect dimension?

### ISSUE THREAD

**Issue by breuderink** [lucidrains/perceiver-pytorch#46]
I am studying multi-head attention. When I was reading through [1], I found that the attenion softmax is applied over the last dimension of the similarity tensor `sim`:

```
        q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> (b h) n d', h = h), (q, k, v))

        sim = einsum('b i d, b j d -> b i j', q, k) * self.scale

        if exists(mask):
            <removed>

        # attention, what we cannot get enough of
        attn = sim.softmax(dim = -1)
 ```

If I understand correctly `sim` has the shape `(b*h) n1 n2`. The softmax is computed over the last dimension `n2`. Shouldn't the softmax be applied to matrices with all the similarity values of a single head (i.e. with shape `n1, n2`)?

[1] https://github.com/lucidrains/perceiver-pytorch/blob/main/perceiver_pytorch/perceiver_io.py#L97

**Comment by lucidrains** [comment]
@breuderink Hi Boris, so it is `n2` only, because you'd like to determine, for each query token (`n1`), the best matching key (`n2`), and route the value from `n2` to it. or at least that is how it is traditionally done. if you do end up trying it your way, i'd be curious about your results :)

**Comment by breuderink** [comment]
@lucidrains That totally makes sense. I think I interpreted equation (1) in https://arxiv.org/pdf/1706.03762.pdf incorrectly. Thanks a lot. It has been a pleasure studying your code.
